README.md

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/knitrProgressBar)](https://cran.r-project.org/package=knitrProgressBar) [![Build Status](https://travis-ci.org/rmflight/knitrProgressBar.svg?branch=master)](https://travis-ci.org/rmflight/knitrProgressBar) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/rmflight/knitrProgressBar?branch=master&svg=true)](https://ci.appveyor.com/project/rmflight/knitrProgressBar) [![Coverage Status](https://img.shields.io/codecov/c/github/rmflight/knitrProgressBar/master.svg)](https://codecov.io/github/rmflight/knitrProgressBar?branch=master)

knitrProgressBar
----------------

[This package](https://rmflight.github.io/knitrProgressBar) supplies a progress bar (shamelessly borrowed from `dplyr::progress_estimated`) that has options to **not** have output captured inside a `knitr` chunk.

### Installation

This package can be installed from CRAN by:

    install.packages("knitrProgressBar")

To install the development version of this package, use `devtools`:

    devtools::install_github("rmflight/knitrProgressBar")

Problem
-------

You want to use `knitr` or `rmarkdown`, but you want to see the progress of a longer running calculation in the chunk. You think that you can just use `dplyr::progress_estimated`. But if you do, all the output from the progress bar will be suppressed ([by design, actually](https://github.com/tidyverse/dplyr/blob/master/R/progress.R#L96)).

Solution
--------

This package has two functions, `progress_estimated`, that creates a `Progress` object that has a connection object associated with it, and `update_progress`, that properly updates the progress object. The output from the progress will be written to **that** connection. This connection will be either `stdout` (default within an R session), `stderr` (default from within `knitr`), or to a log-file.

Examples
--------

None of these are run in this document!

### Setup

``` r
library(knitrProgressBar)

# borrowed from example by @hrbrmstr
arduously_long_nchar <- function(input_var, .pb=NULL) {
  
  update_progress(.pb)
  
  Sys.sleep(0.5)
  
  nchar(input_var)
  
}
```

### Let Function Decide

If you want the object to decide where to put output, do nothing. Just call the `progress_estimated()` function, which uses `make_kpb_output_decisions()`:

``` r
pb <- progress_estimated(length(letters))

purrr::map_int(letters, arduously_long_nchar, .pb = pb)
```

See the help and the vignette for an explanation of how `make_kpb_output_decisions()` decides where to display the progress bar output.

### Write to Specific Connection

If you want to write the progress out to a specific connection, just pass the connection to the `progress_estimated()` call:

``` r
pb <- progress_estimated(length(letters), progress_location = stdout())

purrr::map_int(letters, arduously_long_nchar, .pb = pb)
```

This includes specific files. You can then display the file, or use `tailf` or equivalent to watch the output of the file.

``` r
pb <- progress_estimated(length(letters), progress_location = file("progress.log", open = "w"))

purrr::map_int(letters, arduously_long_nchar, .pb = pb)
```

Connection Considerations
-------------------------

Each connection will display in specific situations, notably `stdout()` will not display to the terminal when run as part of a document being `knit`ted.

Inspiration
-----------

This package (and the examples) was inspired by [this post](https://rud.is/b/2017/03/27/all-in-on-r%E2%81%B4-progress-bars-on-first-post/) from Bob Rudis! Also, thanks to Hadley Wickham for the great `Progress` object and methods!

Website
-------

Web accessible documentation is available [here](https://rmflight.github.io/knitrProgressBar).

Bug Reports
-----------

Please submit bug reports using the GitHub [issue tracker](https://github.com/rmflight/knitrProgressBar/issues).

Code of Conduct
---------------

Please note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.

License
-------

This package is licensed using an [MIT license](LICENSE.md), copyright Robert M Flight.


NEWS.md
# knitrProgressBar 1.1.0

* Added ability to see progress for multi-process taks. See the 
vignette "Progress for Multi-Processing" for more details.

# knitrProgressBar 1.0.0

* Added testing on travis-ci and appveyor.

# knitrProgressBar 0.9.0

* new function for deciding where to write the output from the progress bar, `make_kpb_output_decisions()`. #4. Note that this can be overridden by supplying a connection directly.

* new function for updating the counter easily, `update_progress`, that provides
updating and stopping the counter properly. #7

* Added CI to the github repo. #3

* Added testing. #5



tests/testthat/test-kpb.R
context("kpb")

run_pb <- function(){
  n_iter <- 5
  pb <- progress_estimated(n_iter)
  tmp <- purrr::map(seq(1, n_iter), function(x){
    pb$tick()
    pb$print()
    Sys.sleep(0.1)
  })
}

test_that("regular progress works", {
  with_mock(
    is_interactive = function() TRUE,
    is_in_knitr = function() FALSE,

    {
      pb <- progress_estimated(5)
      withr::with_output_sink("test_pb.txt",
                run_pb())
      expect_match(paste(suppressMessages(readLines("test_pb.txt", warn = FALSE)), collapse = ""), "|========================================================================================================|100% ~0 s remaining")
      file.remove("test_pb.txt")
    }

  )
})

test_that("length warnings and errors work", {
  expect_error(progress_estimated(1), "n must be > 1!")
  expect_warning(progress_estimated(letters), "n is not a number, trying to take the length of n ...")
})

tests/testthat/test-decision-logic.R
context("decision-logic")

test_that("utility functions work", {

  expect_match(get_con_description(stdout()), "stdout")

  expect_false(is_interactive())

  expect_false(is_in_knitr())
})

test_that("logfile choices work",
  with_mock(
    get_chunk_label = function() "knitrchunk",
    is_in_knitr = function() TRUE,
    get_con_description = function(con){
      class(con) <- "connection"
      con_desc <- unlist(summary.connection(con))["description"]
      close.connection(con)
      con_desc
    },

    all_options <- list(kpb.log_file = "testing_1.log"),
    expect_match(get_con_description(set_logfile(all_options)), "testing_1.log"),

    all_options <- list(kpb.log_pattern = "hi"),
    expect_match(get_con_description(set_logfile(all_options)), "hiknitrchunk.log"),

    all_options <- list(),
    expect_match(get_con_description(set_logfile(all_options)), "kpb_output.log")

  )
)


test_that("basic decisions work",
  {with_mock(
    is_interactive = function() TRUE,

    is_in_knitr = function() TRUE,
    expect_match(get_con_description(make_kpb_output_decisions()), "stderr")
  )

  with_mock(
    is_interactive = function() TRUE,
    is_in_knitr = function() FALSE,
    expect_match(get_con_description(make_kpb_output_decisions()), "stdout")
  )

  with_mock(
    is_interactive = function() FALSE,
    is_in_knitr = function() TRUE,
    expect_match(get_con_description(make_kpb_output_decisions()), "stderr")
  )

  with_mock(
    is_interactive = function() FALSE,
    is_in_knitr = function() FALSE,
    expect_match(get_con_description(make_kpb_output_decisions()), "stdout")
  )

  with_mock(
    is_interactive = function() FALSE,
    is_in_knitr = function() FALSE,
    withr::with_options(list(kpb.suppress_noninteractive = TRUE),
                 expect_null(make_kpb_output_decisions()))
  )

  with_mock(
    is_interactive = function() FALSE,
    is_in_knitr = function() TRUE,
    withr::with_options(list(kpb.suppress_noninteractive = TRUE),
                        expect_null(make_kpb_output_decisions()))
  )
  }

)
#
test_that("logfile options work",
  {
    with_mock(
      is_interactive = function() TRUE,
      is_in_knitr = function() FALSE,
      withr::with_options(list(kpb.use_logfile = TRUE),
                          {
                            withr::with_output_sink("logfile_message.txt",
                                                   make_kpb_output_decisions())
                          expect_match(paste(scan("logfile_message.txt", sep = "", what = character(), blank.lines.skip = FALSE), collapse = " "),  " Progress is being logged in: kpb_output.log")
                            }
    ))
    }
  )

R/progress_mp.R
#' multi process progress indicator
#'
#' Sets up a progress object that writes to a shared file to indicate the total
#' progress. Progress can be monitored by `watch_progress_mp`.
#'
#' @param write_location where to save progress to
#'
#' @seealso watch_progress_mp
#' @export
#' @return ProgressMP
set_progress_mp <- function(write_location = NULL){
  ProgressMP$new(write_location = write_location)
}

ProgressMP <- R6::R6Class("ProgressMP",
  inherit = Progress,
  public = list(
    write_location = NULL,
    print = function(){
      if (!is.null(self$write_location)) {
        cat(".", file = self$write_location, append = TRUE)
      } else {
        return(self)
      }
    },
    tick = function(){return(self)},

    initialize = function(write_location = NULL) {
      if (is.null(write_location)) {
        self$write_location <- tempfile(pattern = "kbp_mp_progress", fileext = ".log")
      } else {
        self$write_location <- write_location
      }
      cat("", file = self$write_location, append = FALSE)
      self$i <- 0
      self$n <- 1
    }
  )
)

#' watch progress from multi process
#'
#' sets up a "watcher" function that will report on the progress
#' of a multi-process process that is being indicated by `set_progress_mp`.
#'
#' @param n number of times process is running
#' @param min_time how long to wait
#' @param watch_location where is the progress being written to
#' @param progress_location where to write the progress output
#'
#' @seealso set_progress_mp
#' @export
#' @return ProgressMPWatcher
watch_progress_mp <- function(n, min_time = 0, watch_location = NULL,
                              progress_location = make_kpb_output_decisions()) {
  if (is.null(watch_location)) {
    stop("The watch_location must be set!")
  }
  ProgressMPWatcher$new(n = n, min_time = min_time,
                        watch_location = watch_location,
                        progress_location = progress_location)
}

ProgressMPWatcher <- R6::R6Class("ProgressMPWatcher",
  inherit = Progress,
  public = list(

    watch_location = NULL,
    #self$watch_interval = 0.5,

    initialize = function(n, min_time = 0, watch_location, progress_location) {
      self$progress_location <- progress_location
      self$n <- n
      self$min_time <- min_time

      self$watch_location <- watch_location
      #self$progress_location <- stdout()
      self$begin()
      self$tick()
    },

    tick = function(){
      if (file.exists(self$watch_location)) {
        read_dot <- scan(file = self$watch_location, what = character(), quiet = TRUE)
        if (length(read_dot) > 0) {
          n_dot <- nchar(read_dot)
          if (n_dot > self$i) {
            self$i <- n_dot
            self$print()
          } else if (n_dot == self$n) {
            self$stop()$print()
            return()
          }
        }
      }
      later::later(self$tick, 1)
    }
  )
)


R/decision_logic.R
#' Progress Output Location
#'
#' Provides functionality to decide **how** the progress should be written,
#' if at all.
#'
#' @details
#'
#' This function makes decisions about **how** the progress bar should be displayed
#' based on whether:
#'
#' 1. The code is being run in an interactive session or not
#' 1. The code is part of a `knitr` evaluation using `knit()` or `rmarkdown::render()`
#' 1. Options set by the user. These options include:
#'     1. **kpb.suppress_noninteractive**: a logical value. Whether to suppress output
#'   when being run non-interactively.
#'     1. **kpb.use_logfile**: logical, should a log-file be used for output?
#'     1. **kpb.log_file**: character string defining the log-file to use. **kpb.use_logfile** must be `TRUE`.
#'     1. **kpb.log_pattern**: character string providing a pattern to use, will be combined with the chunk
#'   label to create a log-file for each knitr chunk. **kpb.use_logfile** must be `TRUE`.
#'
#' Based on these, it will either return a newly opened connection, either via
#' `stderr()`, `stdout()`, or a file connection via `file("logfile.log", open = "w")`.
#' Note that for files this will overwrite a previously existing file, and the contents
#' will be lost.
#'
#' @examples
#' \dontrun{
#' # suppress output when not interactive
#' options(kpb.suppress_noninteractive = TRUE)
#'
#' # use a log-file, will default to kpb_output.txt
#' options(kpb.use_logfile = TRUE)
#'
#' # use a specific log-file
#' options(kpb.use_logfile = TRUE)
#' options(kpb.log_file = "progress.txt")
#'
#' # use a log-file based on chunk names
#' options(kpb.use_logfile = TRUE)
#' options(kpb.log_pattern = "pb_out_")
#' # for a document with a chunk labeled: "longcalc", this will generate "pb_out_longcalc.log"
#' }
#'
#' @export
#'
#' @return a write-able connection or NULL
make_kpb_output_decisions <- function(){

  all_options <- options()
  if (is.null(all_options$kpb.suppress_noninteractive)) {
    suppress_noninteractive <- FALSE
  } else {
    suppress_noninteractive <- all_options$kpb.suppress_noninteractive
  }

  if (is.null(all_options$kpb.use_logfile)) {
    use_logfile <- FALSE
  } else {
    use_logfile <- all_options$kpb.use_logfile
  }

  # dont worry about suppressing a non-interactive session or using log-files first

  # use stderr to see progress if we are in knitr and not use a log-file
  if (is_interactive() && is_in_knitr() && !suppress_noninteractive) {
    pb_connection <- stderr()
  } else if (!is_interactive() && is_in_knitr() && !suppress_noninteractive) {
    # regardless of whether interactive or not, because knitr suppresses the output to stdout
    pb_connection <- stderr()
  } else if (is_interactive() && !is_in_knitr() && !suppress_noninteractive) {
    # however, we can use stdout as soon as we are not in knitr itself
    pb_connection <- stdout()
  } else if (!is_interactive() && !is_in_knitr() && !suppress_noninteractive) {
    pb_connection <- stdout()
  } else if (!is_interactive() && !is_in_knitr() && suppress_noninteractive) {
    # now address suppressing non-interactive
    pb_connection <- NULL
  } else if (!is_interactive() && is_in_knitr() && suppress_noninteractive) {
    # now address suppressing non-interactive
    pb_connection <- NULL
  }

  if (use_logfile) {
    log_connection <- set_logfile(all_options)

    if (!is.null(pb_connection)) {
      log_message <- paste0("\nProgress is being logged in: ",
                            get_con_description(log_connection), "\n")
      cat(log_message, file = pb_connection)
    }
    # replace the progress bar connection with our new one, b/c we are pushing
    # it to the log file
    pb_connection <- log_connection
  }

  pb_connection
}

get_con_description <- function(con){
  unlist(summary.connection(con))["description"]
}

# defining our own version of `interactive` so we can mock it in the tests
is_interactive <- function() {interactive()}

is_in_knitr <- function() {
  isTRUE(getOption("knitr.in.progress"))
}

get_chunk_label <- function() {
  if (is_in_knitr()){
    out_label <- knitr::opts_current$get()$label
  } else {
    out_label <- ""
  }
  out_label
}

set_logfile <- function(all_options) {
  if (!is.null(all_options$kpb.log_file)) {
    logfile <- file(all_options$kpb.log_file, open = "w")
    attr(logfile, "class") <- "kpblogfile"
    return(logfile)
  }

  if (is_in_knitr() && !is.null(all_options$kpb.log_pattern)) {
    chunk_label <- get_chunk_label()
    logfile <- file(paste0(all_options$kpb.log_pattern, chunk_label, ".log"), open = "w")
    attr(logfile, "class") <- "kpblogfile"
    return(logfile)
  }

  logfile <- file("kpb_output.log", open = "w")
  attr(logfile, "class") <- "kpblogfile"
  logfile

}



R/progress.R
# The original source for much of this came from Hadley Wickham's dplyr
# code in github.com/tidyverse/dplyr/R/progress.R.
#
# Below is the original license statement for the dplyr package.
#
# The MIT License (MIT)
# =====================
#
#   Copyright © 2013-2015 RStudio and others.
#
# Permission is hereby granted, free of charge, to any person
# obtaining a copy of this software and associated documentation
# files (the “Software”), to deal in the Software without
# restriction, including without limitation the rights to use,
# copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following
# conditions:
#
#   The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
# OTHER DEALINGS IN THE SOFTWARE.
#
#

# changes to these functions by Robert M Flight
# 1. addition of progress_location as a field of the Progress object, argument to
#    initialize, and argument of constructor
# 2. addition of file argument to cat_line function, and all calls to it
# 3. updates to documentation due to new functionality
# 4. removal of decision logic in the Progress$initialize function
# 5. Addition of Progress$finalizer to close file connections that are given

#' Progress bar with estimated time.
#'
#' This provides a reference class representing a text progress bar that displays the
#' estimated time remaining. When finished, it displays the total duration.  The
#' automatic progress bar can be disabled by setting `progress_location = NULL`.
#'
#' @param n Total number of items
#' @param min_time Progress bar will wait until at least `min_time`
#'   seconds have elapsed before displaying any results.
#' @param progress_location where to write the progress to. Default is to make
#' decisions based on location type using `make_kpb_output_decisions()`.
#'
#' @seealso [make_kpb_output_decisions()]
#'
#' @return A ref class with methods `tick()`, `print()`,
#'   `pause()`, and `stop()`.
#' @export
#' @examples
#' p <- progress_estimated(3)
#' p$tick()
#' p$tick()
#' p$tick()
#'
#' p <- progress_estimated(3)
#' for (i in 1:3) p$pause(0.1)$tick()$print()
#'
#' p <- progress_estimated(3)
#' p$tick()$print()$
#'  pause(1)$stop()
#'
#' # If min_time is set, progress bar not shown until that many
#' # seconds have elapsed
#' p <- progress_estimated(3, min_time = 3)
#' for (i in 1:3) p$pause(0.1)$tick()$print()
#'
#' \dontrun{
#' p <- progress_estimated(10, min_time = 3)
#' for (i in 1:10) p$pause(0.5)$tick()$print()
#'
#' # output to stderr
#' p <- progress_estimated(10, progress_location = stderr())
#'
#' # output to a file
#' p <- progress_estimated(10, progress_location = tempfile(fileext = ".log"))
#' }
#'
#' @importFrom R.oo abort
progress_estimated <- function(n, min_time = 0, progress_location = make_kpb_output_decisions()) {
  if (!inherits(n, c("numeric", "integer"))) {
    warning("n is not a number, trying to take the length of n ...")
    n <- length(n)
  }
  if (n <= 1) {
    stop("n must be > 1!")
  }
  Progress$new(n, min_time = min_time, progress_location = progress_location)
}

#' @importFrom R6 R6Class
Progress <- R6::R6Class("Progress",
  public = list(
    n = NULL,
    i = 0,
    init_time = NULL,
    stopped = FALSE,
    stop_time = NULL,
    min_time = NULL,
    last_update = NULL,
    progress_location = NULL,

    initialize = function(n, min_time = 0, progress_location, ...) {
      self$progress_location <- progress_location
      self$n <- n
      self$min_time <- min_time
      self$begin()
    },

    begin = function() {
      "Initialise timer. Call this before beginning timing."
      self$i <- 0
      self$last_update <- self$init_time <- now()
      self$stopped <- FALSE
      self
    },

    pause = function(x) {
      "Sleep for x seconds. Useful for testing."
      Sys.sleep(x)
      self
    },

    width = function() {
      getOption("width") - nchar("|100% ~ 99.9 h remaining") - 2
    },

    tick = function() {
      "Process one element"
      if (self$stopped) return(self)

      if (self$i == self$n) abort("No more ticks")
      self$i <- self$i + 1
      self
    },

    stop = function() {
      if (self$stopped) return(self)

      self$stopped <- TRUE
      self$stop_time <- now()
      self
    },

    print = function(...) {
      if (is.null(self$progress_location)) {
        return(invisible(self))
      }

      if (self$stopped) {
        overall <- show_time(self$stop_time - self$init_time)
        if (self$i == self$n) {
          cat_line(file = self$progress_location, "Completed after ", overall)
          cat("\n")
        } else {
          cat_line("Killed after ", overall)
          cat("\n")
        }
        return(invisible(self))
      }

      now_ <- now()
      if (now_ - self$init_time < self$min_time || now_ - self$last_update < 0.05) {
        return(invisible(self))
      }
      self$last_update <- now_

      avg <- (now() - self$init_time) / self$i
      time_left <- (self$n - self$i) * avg
      nbars <- trunc(self$i / self$n * self$width())

      cat_line(file = self$progress_location,
        "|", str_rep("=", nbars), str_rep(" ", self$width() - nbars), "|",
        format(round(self$i / self$n * 100), width = 3), "% ",
        "~", show_time(time_left), " remaining"
      )

      invisible(self)
    },

    finalize = function() {
      if (!is.null(self$progress_location)) {
        if (inherits(self$progress_location, "kpblogfile")) {
          close.connection(self$progress_location)
        }
      }

    }

  )
)

cat_line <- function(file = NULL, ...) {
  msg <- paste(..., sep = "", collapse = "")
  gap <- max(c(0, getOption("width") - nchar(msg, "width")))
  cat("\r", msg, rep.int(" ", gap), sep = "", file = file, append = TRUE)
  utils::flush.console()
}

str_rep <- function(x, i) {
  paste(rep.int(x, i), collapse = "")
}

show_time <- function(x) {
  if (x < 60) {
    paste(round(x), "s")
  } else if (x < 60 * 60) {
    paste(round(x / 60), "m")
  } else {
    paste(round(x / (60 * 60)), "h")
  }
}

now <- function() proc.time()[[3]]
